### PR TITLE
polymc: init at 1.0.4

### PR DESCRIPTION
**Pull request by starcraft66**
###### Motivation for this change

Upstream maintainers of MultiMC have become increasingly hostile to 3rd party package maintainers (such as myself) ever since the addition of Microsoft account login was added to MultiMC. We've been asked to remove official MultiMC branding from our package in the past so this is a good opportunity to switch the MultiMC package over to the [PolyMC](https://github.com/PolyMC/PolyMC) fork who's mission is to swich away from official MultiMC branding and provide good 3rd party packages.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [x] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/games/polymc/default.nix [hunk 1]
```diff
@@ -1,15 +1,15 @@
-{ lib, mkDerivation, fetchFromGitHub, cmake, jdk8, jdk, zlib, file, makeWrapper, xorg, libpulseaudio, qtbase, libGL, msaClientID ? "" }:
+{ lib, mkDerivation, makeDesktopItem, fetchFromGitHub, cmake, jdk8, jdk, zlib, file, makeWrapper, xorg, libpulseaudio, qtbase, libGL, msaClientID ? "" }:
```
Comment (muscaln) [inline pkgs/games/polymc/default.nix:1]: Could you run nixpkgs-format?

File: pkgs/games/polymc/default.nix [hunk 2]
```diff
@@ -23,34 +23,44 @@ in mkDerivation rec {
...
+    install -Dm644 ../launcher/resources/multimc/scalable/launcher.svg $out/share/pixmaps/polymc.svg
+    install -Dm755 ${desktopItem}/share/applications/polymc.desktop $out/share/applications/polymc.desktop
 
     # xorg.xrandr needed for LWJGL [2.9.2, 3) https://github.com/LWJGL/lwjgl/issues/128
-    wrapProgram $out/bin/multimc \
+    wrapProgram $out/bin/polymc \
       --set GAME_LIBRARY_PATH /run/opengl-driver/lib:${libpath} \
-      --prefix PATH : ${lib.makeBinPath [ xorg.xrandr ]}
+      --prefix PATH : ${lib.makeBinPath [ xorg.xrandr ]} \
+      --add-flags '-d "''${XDG_DATA_HOME-$HOME/.local/share}/polymc"'
```
Comment (muscaln) [inline pkgs/games/polymc/default.nix:53]: PolyMC should work without this flag.
https://github.com/PolyMC/PolyMC/blob/1e51827ba3d986caa9e00871f538107ebe7b1419/launcher/Application.cpp#L307

```suggestion

```

**Review by muscaln — CHANGES_REQUESTED** [review]

File: pkgs/games/polymc/default.nix [hunk 3]
```diff
@@ -23,34 +23,44 @@ in mkDerivation rec {
...
 
   meta = with lib; {
-    homepage = "https://multimc.org/";
+    homepage = "https://polymc.org/";
     description = "A free, open source launcher for Minecraft";
     longDescription = ''
       Allows you to have multiple, separate instances of Minecraft (each with their own mods, texture packs, saves, etc) and helps you manage them and their associated options with a simple interface.
     '';
     platforms = platforms.linux;
     license = licenses.asl20;
```
Comment (lourkeur) [inline pkgs/games/polymc/default.nix:63]:
```suggestion
    license = licenses.gpl3Plus;
```
Per upstream.

File: pkgs/games/polymc/default.nix (hunk 1 above)
Comment (starcraft66) [reply pkgs/games/polymc/default.nix:1]: done

File: pkgs/games/polymc/default.nix (hunk 2 above)
Comment (starcraft66) [reply pkgs/games/polymc/default.nix:53]: My bad, forgot to remove this workaround!

File: pkgs/games/polymc/default.nix (hunk 3 above)
Comment (starcraft66) [reply pkgs/games/polymc/default.nix:63]: Thanks

File: pkgs/top-level/aliases.nix [hunk 4]
```diff
@@ -1195,6 +1195,11 @@ mapAliases ({
     Its new location is obs-studio-plugins.wlrobs.
   '';
 
+  # Added 2022-01-08
+  multimc = throw ''
+    multimc has been replaced with a fork. mutltimc is now polymc.
+  '';
```
Comment (AndersonTorres) [inline pkgs/top-level/aliases.nix:1201]: Why the hell a _multiline_ **one-line** string here? This expression distunes completely from the style of this file.

```suggestion
  multimc = throw "multimc was removed from Nixpkgs; use polymc instead"; # Added 2022-01-08
```

File: pkgs/games/polymc/default.nix [hunk 5]
```diff
@@ -23,34 +24,43 @@ in mkDerivation rec {
...
-      --prefix PATH : ${lib.makeBinPath [ xorg.xrandr ]}
+      --prefix PATH : ${lib.makeBinPath [ xorg.xrandr ]} \
   '';
 
   meta = with lib; {
-    homepage = "https://multimc.org/";
+    homepage = "https://polymc.org/";
     description = "A free, open source launcher for Minecraft";
     longDescription = ''
       Allows you to have multiple, separate instances of Minecraft (each with their own mods, texture packs, saves, etc) and helps you manage them and their associated options with a simple interface.
```
Comment (AndersonTorres) [inline pkgs/games/polymc/default.nix:60]: 80-column limit, please.

File: pkgs/games/polymc/default.nix [hunk 6]
```diff
@@ -23,34 +24,43 @@ in mkDerivation rec {
...
+    exec = "polymc";
+    categories = "Game";
+    terminal = "false";
+  };
 
   postInstall = ''
-    install -Dm644 ../launcher/resources/multimc/scalable/multimc.svg $out/share/pixmaps/multimc.svg
-    install -Dm755 ../launcher/package/linux/multimc.desktop $out/share/applications/multimc.desktop
+    install -Dm644 ../launcher/resources/multimc/scalable/launcher.svg $out/share/pixmaps/polymc.svg
+    install -Dm755 ${desktopItem}/share/applications/polymc.desktop $out/share/applications/polymc.desktop
```
Comment (AndersonTorres) [inline pkgs/games/polymc/default.nix:48]: Use the `DesktopItems` trio instead of a custom copy.

File: pkgs/games/polymc/default.nix [hunk 7]
```diff
@@ -1,15 +1,16 @@
...
 let
   libpath = with xorg; lib.makeLibraryPath [ libX11 libXext libXcursor libXrandr libXxf86vm libpulseaudio libGL ];
-in mkDerivation rec {
-  pname = "multimc";
-  version = "unstable-2021-09-08";
+in
+mkDerivation rec {
+  pname = "polymc";
+  version = "1.0.4";
   src = fetchFromGitHub {
```
Comment (AndersonTorres) [inline pkgs/games/polymc/default.nix:9]:
```suggestion

  src = fetchFromGitHub {
```

File: pkgs/games/polymc/default.nix [hunk 8]
```diff
@@ -1,15 +1,16 @@
...
-    repo = "MultiMC5";
-    rev = "e2355eb276bf355ca4acf526a0f3cc390aa88f8b";
-    sha256 = "3G9QPoAbC+uVfUYR0Kq6hnxl9c2mvCzIEYGjwfarQJ8=";
+    owner = "PolyMC";
+    repo = "PolyMC";
+    rev = version;
+    sha256 = "sha256-8aya0KfV9F+i2qBpweWcR9hwyTSQkqn2wHdtkCEeNvk=";
     fetchSubmodules = true;
   };
   nativeBuildInputs = [ cmake file makeWrapper ];
```
Comment (AndersonTorres) [inline pkgs/games/polymc/default.nix:16]:
```suggestion

  nativeBuildInputs = [ cmake file makeWrapper ];
```

File: pkgs/games/polymc/default.nix [hunk 9]
```diff
@@ -23,34 +24,43 @@ in mkDerivation rec {
...
-    substituteInPlace notsecrets/Secrets.cpp \
-      --replace 'QString MSAClientID = "";' 'QString MSAClientID = "${msaClientID}";'
+    substituteInPlace CMakeLists.txt \
+      --replace '17b47edd-c884-4997-926d-9e7f9a6b4647' '${msaClientID}'
   '';
 
-  cmakeFlags = [ "-DMultiMC_LAYOUT=lin-system" ];
+  cmakeFlags = [ "-DLauncher_LAYOUT=lin-system" ];
+
+  desktopItem = makeDesktopItem {
```
Comment (AndersonTorres) [inline pkgs/games/polymc/default.nix:35]: Use the DesktopItems trio instead of a custom copy.

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/games/polymc/default.nix (hunk 9 above)
Comment (starcraft66) [reply pkgs/games/polymc/default.nix]: done

File: pkgs/games/polymc/default.nix (hunk 8 above)
Comment (starcraft66) [reply pkgs/games/polymc/default.nix:18]: done

File: pkgs/games/polymc/default.nix (hunk 7 above)
Comment (starcraft66) [reply pkgs/games/polymc/default.nix:10]: done

File: pkgs/games/polymc/default.nix (hunk 6 above)
Comment (starcraft66) [reply pkgs/games/polymc/default.nix]: done

File: pkgs/games/polymc/default.nix (hunk 5 above)
Comment (starcraft66) [reply pkgs/games/polymc/default.nix]: shortened to <80 cols

File: pkgs/top-level/aliases.nix (hunk 4 above)
Comment (starcraft66) [reply pkgs/top-level/aliases.nix]: fixed

**Comment by thelegy** [comment]
Can we insert the msa client-id in the buildprocess without referencing to the upstream client-id? Maybe with some kind of regex replacement actually referencing the field name and not its upstream value?

I am unsure if microsofts terms and services would allow for that: they treat the msa client-ids as secrets. If polymc decides to bake theirs into their repo it is their decision but I feel we should not needlessly replicate that attitude.

A more practical reasoning might be, that replacing a particular client-id will break (probably without any errors and thus no one noticing) when the upstream value changes.

**Comment by erdnaxe** [comment]
> Can we insert the msa client-id in the buildprocess without referencing to the upstream client-id? Maybe with some kind of regex replacement actually referencing the field name and not its upstream value?

From what I understand, `Launcher_MSA_CLIENT_ID` can be passed as a CMake option directory because https://github.com/PolyMC/PolyMC/blob/develop/CMakeLists.txt#L78 does not change the value if it's already in the cache.

File: pkgs/games/polymc/default.nix [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+      desktopName = "PolyMC";
+      genericName = "Minecraft Launcher";
+      comment = "Free, open source launcher and instance manager for Minecraft.";
+      icon = "launcher";
+      exec = "polymc";
+      categories = "Game";
+      terminal = "false";
+    })
+  ];
+
```
Comment (muscaln) [inline pkgs/games/polymc/default.nix:49]:
```suggestion

dontWrapQtApps = true;

```

File: pkgs/games/polymc/default.nix [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+  postInstall = ''
+    install -Dm644 ../launcher/resources/multimc/scalable/launcher.svg $out/share/pixmaps/polymc.svg
+
+    # xorg.xrandr needed for LWJGL [2.9.2, 3) https://github.com/LWJGL/lwjgl/issues/128
+    wrapProgram $out/bin/polymc \
+      --set GAME_LIBRARY_PATH /run/opengl-driver/lib:${libpath} \
+      --prefix PATH : ${lib.makeBinPath [ xorg.xrandr ]} \
```
Comment (muscaln) [inline pkgs/games/polymc/default.nix:56]:
```suggestion
    wrapProgram $out/bin/polymc \
      "''${qtWrapperArgs[@]}" \
      --set GAME_LIBRARY_PATH /run/opengl-driver/lib:${libpath} \
      --prefix PATH : ${lib.makeBinPath [ xorg.xrandr ]}
```

**Review by muscaln — CHANGES_REQUESTED** [review]
It looks like we are double-wrapping the binary.

File: pkgs/games/polymc/default.nix [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+      genericName = "Minecraft Launcher";
+      comment = "Free, open source launcher and instance manager for Minecraft.";
+      icon = "launcher";
+      exec = "polymc";
+      categories = "Game";
+      terminal = "false";
+    })
+  ];
+
+  postInstall = ''
```
Comment (AndersonTorres) [inline pkgs/games/polymc/default.nix:50]: It is better to put the `let libpath =... ; in ` here, nearer to its intended use.

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/games/polymc/default.nix (hunk 12 above)
Comment (starcraft66) [reply pkgs/games/polymc/default.nix]: Made it more readable

File: pkgs/games/polymc/default.nix (hunk 11 above)
Comment (starcraft66) [reply pkgs/games/polymc/default.nix]: fixed.

**Comment by starcraft66** [comment]
@erdnaxe I removed the client ID and inserted it as a cmake build flag

**Review by AndersonTorres — APPROVED** [review]

**Review by erdnaxe — APPROVED** [review]

**Comment by AndersonTorres** [comment]
@starcraft66 please fix the conflicts.

Further, I suggest putting the documentation updates in a separate commit.

(I don't remember correctly, but is the XML file machine-generated?)

**Comment by starcraft66** [comment]
I rebased and split the commit as requested.

File: pkgs/games/polymc/default.nix [hunk 13]
```diff
@@ -0,0 +1,77 @@
+{ lib, mkDerivation, makeDesktopItem, fetchFromGitHub, cmake, jdk8, jdk, zlib, file, makeWrapper, xorg, libpulseaudio, qtbase, libGL, msaClientID ? "" }:
```
Comment (AndersonTorres) [inline pkgs/games/polymc/default.nix:1]: This line is huge. Can you put one parameter per line? 

(Or at least break it in the 80-char rule...)

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/games/polymc/default.nix (hunk 13 above)
Comment (starcraft66) [reply pkgs/games/polymc/default.nix]: done

**Comment by AndersonTorres** [comment]
LGTM. Waiting for OfBorg.

**Comment by lourkeur** [comment]
poke @AndersonTorres

**Comment by AndersonTorres** [comment]
Ping @starcraft66 to fix merge conflicts.

**Comment by NIXFOREVERNIX** [comment]
polymc still need overide myself msaClientID ?

**Comment by starcraft66** [comment]
@AndersonTorres done

@NIXFOREVERNIX You won't need to override it as PolyMC provides one by default but you can if you prefer to.

**Comment by AndersonTorres** [comment]
_Muchas gracias!_

**Comment by NIXFOREVERNIX** [comment]
@starcraft66 Thank you for your work!

**Comment by NIXFOREVERNIX** [comment]
@starcraft66 .desktop file **exec**  parameter empty
```console
[Desktop Entry]
Version=1.0
Name=PolyMC
Comment=A custom launcher for Minecraft that allows you to easily manage multiple installations of Minecraft at once.
Type=Application
Terminal=false

Exec=

StartupNotify=true
Icon=org.polymc.PolyMC
Categories=Game;
Keywords=game;minecraft;launcher;
```

**Comment by starcraft66** [comment]
Okay, I see what's going on here. I think the desktop file you're looking at comes from the PolyMC source repo and isn't the one from this package. Admittedly, I don't use desktop launchers and didn't test this myself which I should do in the future.

I'll take a look at this and open a PR this week to fix it.

**Comment by NIXFOREVERNIX** [comment]
> Okay, I see what's going on here. I think the desktop file you're looking at comes from the PolyMC source repo and isn't the one from this package. Admittedly, I don't use desktop launchers and didn't test this myself which I should do in the future.
> 
> I'll take a look at this and open a PR this week to fix it.

Thank you, 
The desktop file in **/nix/store/0xn1rrlz6fj2cwjzfb3pqh4s3mig4q50-polymc-1.0.4/share/applications/org.polymc.PolyMC.desktop**

This is **github:PolyMC/program_info/org.polymc.PolyMC.desktop.in** file:
```console
...
Exec=@Launcher_APP_BINARY_NAME@
...
```
Maybe can't read "Launcher_APP_BINARY_NAME" , But i don't konw how to fix.

**Pull request merged**